### PR TITLE
[native pos] Do not create empty broadcast files 

### PR DESCRIPTION
**Pull request by singcha**
Fix the issue with empty broadcast files by skipping file generation if there is nothing to write. Do not return file name in result if no rows are written for broadcast. This removes any handling for empty files on read path.

Test plan - Added unit test

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp [hunk 1]
```diff
@@ -146,10 +146,16 @@ velox::BufferPtr BroadcastFileReader::next() {
 
   auto readFile = fileSystem_->openFileForRead(
       broadcastFileInfos_[readfileIndex_].filePath_);
-  auto buffer = AlignedBuffer::allocate<char>(readFile->size(), pool_, 0);
-  readFile->pread(0, readFile->size(), buffer->asMutable<char>());
   ++readfileIndex_;
   ++numFiles_;
+
+  // Empty broadcast file.
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp:152]: Why would we produce an empty broadcast file in the first place?

Comment (singcha) [reply presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp:152]: This gets generated when we have empty vectors. We can add handling to track how many rows have been written and return empty pages/no page for those scenarios.

Comment (mbasmanova) [reply presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp:152]: > We can add handling to track how many rows have been written and return empty pages/no page for those scenarios.

That would be better I think. Thanks.

**Comment by singcha** [comment]
@bot kick off tests

File: presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp (hunk 1 above)
Comment (singcha) [reply presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp:169]: Updated code to not open file till we have non zero rows to write.

File: presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp [hunk 2]
```diff
@@ -78,12 +76,13 @@ BroadcastInfo::BroadcastInfo(
...
 BroadcastFileWriter::BroadcastFileWriter(
-    std::unique_ptr<WriteFile> writeFile,
     std::string_view filename,
-    velox::memory::MemoryPool* pool,
-    const RowTypePtr& inputType)
-    : writeFile_(std::move(writeFile)),
+    const RowTypePtr& inputType,
+    std::shared_ptr<velox::filesystems::FileSystem> fileSystem,
+    velox::memory::MemoryPool* pool)
+    : fileSystem_(fileSystem),
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp:83]: std::move

File: presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp [hunk 3]
```diff
@@ -96,6 +95,11 @@ void BroadcastFileWriter::noMoreData() {}
 
 // TODO: Add file stats - size, checksum, number of rows.
 RowVectorPtr BroadcastFileWriter::fileStats() {
+  // No rows written.
+  if (numRows_ == 0) {
+    return nullptr;
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp:100]: It would be nice to update comments for this method to explain that it may return nullptr.

File: presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp [hunk 4]
```diff
@@ -146,10 +163,16 @@ velox::BufferPtr BroadcastFileReader::next() {
 
   auto readFile = fileSystem_->openFileForRead(
       broadcastFileInfos_[readfileIndex_].filePath_);
-  auto buffer = AlignedBuffer::allocate<char>(readFile->size(), pool_, 0);
-  readFile->pread(0, readFile->size(), buffer->asMutable<char>());
   ++readfileIndex_;
   ++numFiles_;
+
+  // Empty broadcast file.
+  if (readFile->size() <= 0) {
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp:170]: Shouldn't think be a check? It is not possible to have an empty file here, right?

File: presto-native-execution/presto_cpp/main/operators/BroadcastExchangeSource.cpp [hunk 5]
```diff
@@ -39,18 +39,19 @@ void BroadcastExchangeSource::request() {
       return;
     }
 
-    if (!reader_->hasNext()) {
+    auto buffer = reader_->next();
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/operators/BroadcastExchangeSource.cpp:42]: Are these changes needed?

**Review by mbasmanova — COMMENTED** [review]
It would be nice to update commit message and PR title to clarify that we no longer create empty files, hence, nothing to handle on the read path.

File: presto-native-execution/presto_cpp/main/operators/BroadcastExchangeSource.cpp (hunk 5 above)
Comment (singcha) [reply presto-native-execution/presto_cpp/main/operators/BroadcastExchangeSource.cpp:42]: Not anymore

File: presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp (hunk 3 above)
Comment (singcha) [reply presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp:100]: updated

File: presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp (hunk 4 above)
Comment (singcha) [reply presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp:170]: correct, removed this check

**Comment by singcha** [comment]
> It would be nice to update commit message and PR title to clarify that we no longer create empty files, hence, nothing to handle on the read path.

updated

File: presto-native-execution/presto_cpp/main/operators/BroadcastFactory.h [hunk 6]
```diff
@@ -59,13 +59,18 @@ class BroadcastFileWriter {
   /// Returns file name.
   velox::RowVectorPtr fileStats();
 
+  // Initializes write file.
+  void initializeWriteFile();
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/operators/BroadcastFactory.h:63]: Can this method be private?

File: presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp [hunk 7]
```diff
@@ -95,7 +94,13 @@ void BroadcastFileWriter::collect(const RowVectorPtr& input) {
 void BroadcastFileWriter::noMoreData() {}
 
 // TODO: Add file stats - size, checksum, number of rows.
+/// This method returns nullptr if there were no rows written.
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp:97]: Move comments to .h file and remove "This method".

**Review by mbasmanova — COMMENTED** [review]
Looks good % a couple minor comments. Thanks.

File: presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp [hunk 8]
```diff
@@ -106,8 +111,21 @@ RowVectorPtr BroadcastFileWriter::fileStats() {
...
+  }
+}
+
 void BroadcastFileWriter::serialize(const RowVectorPtr& rowVector) {
   auto numRows = rowVector->size();
+  if (numRows == 0) {
+    return;
+  }
+
+  initializeWriteFile();
```
Comment (vermapratyush) [inline presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp:126]: Initiating write file in a serialize method seems odd.
Better to call this method `write`.

**Review by vermapratyush — COMMENTED** [review]
What's the reason to prefer PrestoVectorSerde over UnsafeRowSerde ?

**Review by mbasmanova — APPROVED** [review]
Thanks.

File: presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp (hunk 8 above)
Comment (singcha) [reply presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp:126]: good point! updated

**Comment by singcha** [comment]
> 

Using UnsafeRow is more expensive as this will convert to row format, while with PrestoVectorSerde we keep the data in columnar format.

**Comment by singcha** [comment]
> What's the reason to prefer PrestoVectorSerde over UnsafeRowSerde ?

Using UnsafeRow is more expensive as this will convert to row format, while with PrestoVectorSerde we keep the data in columnar format.
